### PR TITLE
Add support for `markdown` format output

### PR DESCRIPTION
**Pull request by sandrobonazzola**
Add markdown format output and enable it for bugzilla, github and gitlab.

### CODE REVIEW

File: did/cli.py [hunk 1]
```diff
@@ -149,8 +149,15 @@ def parse(self):
             raise RuntimeError(
                 "Invalid date range ({0} to {1})".format(
                     opt.since, opt.until.date - delta(days=1)))
-        header = "Status report for {0} ({1} to {2}).".format(
+
+        header = "Status report for {0} ({1} to {2})".format(
             period, opt.since, opt.until.date - delta(days=1))
+        if opt.format == "md":
+            # in markdown first line must be an header
+            header = "# " + header
```
Comment (kwk) [inline did/cli.py:157]: I'd suggest to use the alternate syntax for headings as described here: https://www.markdownguide.org/basic-syntax/#alternate-syntax. Then you can use `did`'s output in commit messages as well. The `#` headers are always consumed as comments which is a bit unfortunate.

File: did/utils.py [hunk 2]
```diff
@@ -185,7 +185,7 @@ def item(text, level=0, options=None):
     if level == 1 and options.brief:
         return
     # Four space for each level, additional space for wiki format
-    indent = level * 4
+    indent = level * (4 if options.format != "md" else 2)
```
Comment (kwk) [inline did/utils.py:188]: If I'm interpreting this correctly you want to have `2` spaces for the markdown format. I suggest to drop the syntactic sugar and just make it more obvious what's the default and what' special for markdown.

```suggestion
    indent = level * 4
    if options.format == "md":
        indent = level * 2
```

**Review by kwk — COMMENTED** [review]
Niced idea. I have some just some minor thoughts.

File: did/cli.py (hunk 1 above)
Comment (sandrobonazzola) [reply did/cli.py]: done

File: README.rst [hunk 3]
```diff
@@ -86,12 +86,13 @@ The default output is plain text of maximum width 79 characters.
 This can be adjusted using the ``--width`` parameter. To disable
 shortening altogether use ``--width=0``. The default width value
 can be saved in the config file as well. Use ``--format=wiki`` to
-enable simple MoinMoin wiki syntax. For stats which support them,
+enable simple MoinMoin wiki syntax or ``--format=wiki`` to enable
```
Comment (psss) [inline README.rst:89]:
```suggestion
enable simple MoinMoin wiki syntax or ``--format=md`` to enable
```

File: README.rst [hunk 4]
```diff
@@ -86,12 +86,13 @@ The default output is plain text of maximum width 79 characters.
...
+markdown syntax. For stats which support them,
 ``--brief`` and ``--verbose`` can be used to specify a different
 level of detail to be shown.
 
---format=FMT
-    Output style, possible values: text (default) or wiki
+--format {text,md,wiki}
```
Comment (psss) [inline README.rst:94]:
```suggestion
--format {text|md|wiki}
```

Perhaps, `|` is more commonly used to separate alternative values?

File: README.rst (hunk 3 above)
Comment (psss) [reply README.rst:89]: Or, perhaps, would you mind if we use `--format=markdown` instead of the abbreviation?

File: did/plugins/gitlab.py [hunk 5]
```diff
@@ -176,9 +177,18 @@ def iid(self):
...
         """ String representation """
-        return "{0}#{1} - {2}".format(
-            self.project['path_with_namespace'],
-            str(self.id).zfill(PADDING), self.title)
+        if self.parent.options.format == "md":
+            return "[{1}#{3}]({0}/{1}/-/{2}/{3}) - {4}".format(
+                self.gitlabapi.url,
+                self.project['path_with_namespace'],
+                ("issues"
+                    if self.data['target_type'] == 'Issue'
```
Comment (psss) [inline did/plugins/gitlab.py:185]: I tried this with my account and for this issue:

* https://gitlab.com/testing-farm/general/-/issues/41

I got a link to a (non-existent) merge request instead:

* https://gitlab.com/testing-farm/general/-/merge_requests/41

Quickly checked the `self.data` and I see this:

    'noteable_type': 'Issue'
    'target_type': 'Note'

Perhaps, we should use `noteable_type` here instead?

File: did/plugins/gitlab.py [hunk 6]
```diff
@@ -223,7 +233,7 @@ def fetch(self):
             self.user.login, self.options.since, self.options.until,
             'Issue', 'opened')
         self.stats = [
-            Issue(issue, self.parent.gitlab)
+            Issue(issue, self.parent.gitlab, self.parent)
```
Comment (psss) [inline did/plugins/gitlab.py:236]: So if we hand over the whole `parent` now, we could extract the `self.parent.gitlab` in the constructor as well?

File: did/cli.py [hunk 7]
```diff
@@ -149,8 +149,17 @@ def parse(self):
             raise RuntimeError(
                 "Invalid date range ({0} to {1})".format(
                     opt.since, opt.until.date - delta(days=1)))
-        header = "Status report for {0} ({1} to {2}).".format(
+
+        header = "Status report for {0} ({1} to {2})".format(
             period, opt.since, opt.until.date - delta(days=1))
+        if opt.format == "md":
+            # in markdown first line must be an header
```
Comment (psss) [inline did/cli.py:156]:
```suggestion
            # In markdown the first line must be a header
```

File: did/cli.py [hunk 8]
```diff
@@ -149,8 +149,17 @@ def parse(self):
...
+            # using alternate syntax allowing to use did's
+            # output in commit messages as well
+            header = f"{header}\n{'=' * len(header)}"
+        else:
+            # in markdown no trailing punctuation is allowed in headings
```
Comment (psss) [inline did/cli.py:161]:
```suggestion
            # In markdown no trailing punctuation is allowed in headings
```

Just consistent comments...

**Review by psss — CHANGES_REQUESTED** [review]
Nice improvement! Thanks for implementing this. Added just a few, rather minor, comments.

File: README.rst (hunk 4 above)
Comment (sandrobonazzola) [reply README.rst:94]: This is the separator used by `argparse` in the `--help` output. I've no problem with your suggestion for the README if you're ok it won't match the help output. Let me know what you prefer.

File: README.rst (hunk 3 above)
Comment (sandrobonazzola) [reply README.rst:89]: ok, I'll send a patch

File: did/plugins/gitlab.py (hunk 6 above)
Comment (sandrobonazzola) [reply did/plugins/gitlab.py:236]: yes, I'll change that.

File: did/plugins/gitlab.py (hunk 5 above)
Comment (sandrobonazzola) [reply did/plugins/gitlab.py:185]: Sorry, missed a case, fixing.

File: README.rst (hunk 3 above)
Comment (psss) [inline README.rst:89]:
```suggestion
enable simple MoinMoin wiki syntax or ``--format=markdown`` to enable
```

Last remnant.

**Comment by psss** [comment]
One more thing: It seems that GitHub pull requests are created with url leading to issues:

    did --format markdown --width 0 --github-pull-requests-created

The output looks like this:

Status report for the week 36 (2023-09-04 to 2023-09-10)
========================================================

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Petr Šplíchal <psplicha@redhat.com>
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

* Pull requests created on github: 5
  * [psss/did#320](https://github.com/psss/did/issues/320) - Move line length recommendation to proper section
  * [teemtee/tmt#2315](https://github.com/teemtee/tmt/issues/2315) - Make the upload to PyPI working again
  * [teemtee/tmt#2314](https://github.com/teemtee/tmt/issues/2314) - Adjust test coverage for deep beakerlib libraries
  * [teemtee/tmt#2312](https://github.com/teemtee/tmt/issues/2312) - Release tmt-1.27.0
  * [teemtee/tmt#2308](https://github.com/teemtee/tmt/issues/2308) - Disable `systemd-resolved` to prevent dns failures

GitHub automatically redirects these to the proper url, so it's not that important to get fixed, but probably would be a bit nicer/cleaner. @sandrobonazzola, what do you think?

**Comment by sandrobonazzola** [comment]
> GitHub automatically redirects these to the proper url, so it's not that important to get fixed, but probably would be a bit nicer/cleaner. @sandrobonazzola, what do you think?

I've kind of cheated there to keep the code simpler but yes, some more parsing can be done to get avoid the redirect. I would consider doing it in another PR as enhancement when I'll have some more time to dedicate to this.

**Comment by sandrobonazzola** [comment]
Nevermind, found a more useful way to do it, pushing a new commit

**Comment by psss** [comment]
> Nevermind, found a more useful way to do it, pushing a new commit

Perfect, thanks!

**Review by psss — APPROVED** [review]
Thanks for addressing all comments, added just a tiny adjustment in 2c505c0.

**Comment by sandrobonazzola** [comment]
> Thanks for addressing all comments, added just a tiny adjustment in [2c505c0](https://github.com/psss/did/commit/2c505c0600fbf3adcdec1e207ed9f45ee462f6b1).

LGTM, thanks

**Comment by psss** [comment]
For some reason, the `--format` option is incorrectly render in the docs:

image

Turning off `smartquotes` in 273d33c should fix it.

**Pull request merged**